### ISSUE TITLE
mention scoping

### ISSUE THREAD

**Issue by HeyDude378** [MicrosoftDocs/PowerShell-Docs-Modules#17]
It doesn't seem that variables have to be used _more_ than their assignments; they can also be used _the same number of times_.  You should clarify the language by saying something like "greater than or equal to their assignments".

You should also mention that the variable must be used _within the same scope in which it was declared_ or else it won't be considered to be "used".

I have an example...

**$foo | for-eachobject {if($_ -eq $false){$bar = $true}}

if($bar){write-host "You are still going to get a PSUseDeclaredVarsMoreThanAssignments error even though you are using the bar variable here."}**

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e77c7bc9-9fda-5e82-def1-d1aaa0f62d0a
* Version Independent ID: 33d40987-b49b-186f-1daa-d54a2e76494c
* Content: [UseDeclaredVarsMoreThanAssignments - PowerShell](https://docs.microsoft.com/en-us/powershell/utility-modules/psscriptanalyzer/rules/UseDeclaredVarsMoreThanAssignments?view=ps-modules)
* Content Source: [reference/docs-conceptual/PSScriptAnalyzer/Rules/UseDeclaredVarsMoreThanAssignments.md](https://github.com/MicrosoftDocs/PowerShell-Docs-Modules/blob/main/reference/docs-conceptual/PSScriptAnalyzer/Rules/UseDeclaredVarsMoreThanAssignments.md)
* Product: **powershell**
* Technology: **powershell-conceptual**
* GitHub Login: @sdwheeler
* Microsoft Alias: **sewhee**

**Comment by sdwheeler** [comment]
@bergmeister Can you comment on this issue?

**Comment by bergmeister** [comment]
@sdwheeler The rule's analysis is limited when it comes to variables definition and usage between different scopes. The author is correct that their example produces a false positive but there are also examples where the rule has some intelligence/knowledge between scopes such as in `'$array=@(); $list | ForEach-Object { $array += 'foo' }' `.
This has been a long-standing and well known limitation of the rule. The rule was improved a little bit but due to technical reasons and PowerShell's dynamic scoping, there is a limit to the intelligence of the script analysis when it comes to variable usage. People therefore need to be conscious that any analyzer will bring up false positives and that they can be suppressed.

Because the rule name is used in PSSA settings files, a rename would be quite disruptive and even if we created an alias name I am not sure if this would just make things confusing? The name of the rule should be read as 'You should use your declared variables and not just assign them a value' but I agree as well that the rule name doesn't read very fluently.

**Comment by HeyDude378** [comment]
Hi, thank you for your reply.  I think my first suggested change would be in the description line, "Generally variables that are not used more than their assignments are considered wasteful and not needed."  I would rephrase it to "Generally variables that are not used are considered wasteful and not needed".

I think my second suggested change is to add a "How to Fix" section like you have for some of the other Problems documentation (like UseLiteralInitializerForHashtable) where you mention that "a variable's usage is sometimes ignored if it is not used in the same scope in which it was declared.  Try using it in the same scope."  Something like that.
